src/plugins/gs-plugin-external-apps.c

```diff
@@ -0,0 +1,804 @@
+
+	/* We cache all runtimes because an external runtime may have been
+	 * adopted by the flatpak plugins */
+	if (app_is_flatpak (app) &&
+	    g_strcmp0 (gs_app_get_flatpak_kind_as_str (app), "runtime") == 0) {
+		gs_plugin_cache_add (plugin, gs_app_get_id (app), app);
+		g_debug ("Caching remote '%s'", gs_app_get_id (app));
+	}
+
+	/* only process this app if was created by this plugin */
+	if (g_strcmp0 (gs_app_get_management_plugin (app),
+		       gs_plugin_get_name (plugin)) != 0)
+		return TRUE;
+
+	g_debug ("Refining external app %s", gs_app_get_id (app));
+
+	ext_runtime = gs_plugin_get_app_external_runtime (plugin, app);
+
+	if (!ext_runtime) {
+		g_debug ("Could not understand the asset from the "
+			 "metadata in "
+			 "app %s: %s",
+			 gs_app_get_id_no_prefix (app), metadata);
+		return TRUE;
+	}
+
+	priv = gs_plugin_get_data (plugin);
+
+	if (!gs_flatpak_refine_app (priv->flatpak, app, flags, cancellable, 
+				    error)) {
+		g_debug ("Refining app %s failed!", gs_app_get_id (app));
+		return FALSE;
+	}
+
+	/* We set the state to available because we assume that we can build
+	 * the runtime */
+	if (gs_app_get_state (ext_runtime) != AS_APP_STATE_INSTALLED) {
+		gs_app_set_state (app, AS_APP_STATE_UNKNOWN);
+		gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+	}
+
+	return TRUE;
+}
+
+gboolean
+gs_plugin_add_installed (GsPlugin *plugin,
+			 GsAppList *list,
+			 GCancellable *cancellable,
+			 GError **error)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	return gs_flatpak_add_installed (priv->flatpak, list, cancellable,
+					 error);
+}
+
+gboolean
+gs_plugin_app_install (GsPlugin *plugin,
+		       GsApp *app,
+		       GCancellable *cancellable,
+		       GError **error)
+{
+	g_autofree char *runtime = NULL;
+	g_autofree char *url = NULL;
+	const char *remote_name;
+	GError *internal_error = NULL;
+	GsApp *ext_runtime;
+	GsPluginData *priv;
+
+	/* only process this app if was created by this plugin */
+	if (g_strcmp0 (gs_app_get_management_plugin (app),
+		       gs_plugin_get_name (plugin)) != 0)
+		return TRUE;
+
+	priv = gs_plugin_get_data (plugin);
+
+	ext_runtime = gs_plugin_get_app_external_runtime (plugin, app);
+
+	if (!ext_runtime) {
+		g_debug ("External app '%s' didn't have any asset! "
+			 "Not installing and marking as state unknown!",
+			 gs_app_get_id (app));
+		gs_app_set_state (app, AS_APP_STATE_UNKNOWN);
+
+		return TRUE;
+	}
+
+	gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+
+	if (gs_app_get_state (ext_runtime) == AS_APP_STATE_UNKNOWN) {
+		if (!build_runtime (plugin, app, ext_runtime, error)) {
+			g_debug ("Failed to build runtime '%s'", runtime);
+			return FALSE;
+		}
+
+		if (!gs_flatpak_refine_app (priv->flatpak, ext_runtime,
+					    GS_PLUGIN_REFINE_FLAGS_DEFAULT,
+					    cancellable, error)) {
+			g_debug ("Failed to refine '%s'",
+				 gs_app_get_id (ext_runtime));
+			return FALSE;
+		}
+	}
+
+	if (gs_app_get_state (ext_runtime) == AS_APP_STATE_UNKNOWN) {
```

Comment:
I had this as a defensive block before (because if there's some unaccounted issue that let *build_runtime* pass, then *gs_flatpak_refine_app* can actually succeed while the state is still left as *UNKNOWN*).
However, I realize now that this is not necessary because we also deal with the case after the switch under this.
Will remove. Thanks!